File: Travel.cs
using System.Globalization;
using System.Net;
using System.Net.Http;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;

namespace Glance;

/// One way to get there. Depart is when to walk out of the door (TfL counts the walk to the stop).
public record Journey(DateTime Depart, DateTime Arrive, string Via);

/// Travel events ("Travel: to Office") and their public transport times. Rules in docs/travel.md; UI-free so the tests can run it.
public static class Travel
{
    public static bool Is(Ev e) => !e.AllDay && e.Title.TrimStart().StartsWith("travel", StringComparison.OrdinalIgnoreCase);

    /// "Travel: to Office" → ("office", false), "Travel: from Gym" → ("gym", true). Work and office are one place.
    public static (string Name, bool From) Name(string title)
    {
        var t = Regex.Replace(title.Trim(), @"^travel\s*:?\s*", "", RegexOptions.IgnoreCase);
        var m = Regex.Match(t, @"^(to|from)\s+", RegexOptions.IgnoreCase);
        var name = t[m.Length..].Trim().ToLowerInvariant();
        return (name == "work" ? "office" : name, m.Success && m.Groups[1].Value.Equals("from", StringComparison.OrdinalIgnoreCase));
    }

    /// Named places learned from travel events that have a location (the latest wins), then [extra] from settings on top.
    public static Dictionary<string, string> Places(IEnumerable<Ev> events, IDictionary<string, string>? extra)
    {
        var p = new Dictionary<string, string>();
        foreach (var e in events.Where(e => Is(e) && !string.IsNullOrWhiteSpace(e.Location)).OrderBy(e => e.Start))
            if (Name(e.Title) is (var n, false) && n != "") p[n] = e.Location!;
        foreach (var (k, v) in extra ?? new Dictionary<string, string>()) p[Name(k).Name] = v;
        return p;
    }

    public static string? Destination(Ev t, IReadOnlyList<Ev> events, IReadOnlyDictionary<string, string> places)
    {
        if (!string.IsNullOrWhiteSpace(t.Location)) return t.Location;
        var (name, from) = Name(t.Title);
        if (!from && places.TryGetValue(name, out var p)) return p;
        if (from)   // back to wherever the trip out there started
        {
            var outbound = events.Where(e => Is(e) && e.Start.Date == t.Start.Date && e.End <= t.Start && Name(e.Title) == (name, false)).MaxBy(e => e.Start);
            if (outbound != null && Origin(outbound, events, places) is string o) return o;
        }
        return events.Where(e => !Is(e) && !e.AllDay && e.Start >= t.Start && e.Start <= t.End.AddHours(2) && !string.IsNullOrWhiteSpace(e.Location))
            .MinBy(e => e.Start)?.Location;
    }

    /// Where the trip starts, from the calendar alone: "from X", else the last place you were that day, else home.
    public static string? Origin(Ev t, IReadOnlyList<Ev> events, IReadOnlyDictionary<string, string> places)
    {
        var (name, from) = Name(t.Title);
        if (from && places.TryGetValue(name, out var p)) return p;
        foreach (var e in events.Where(e => e != t && !e.AllDay && e.Start.Date == t.Start.Date && e.End <= t.Start).OrderByDescending(e => e.End))
            if ((Is(e) ? Destination(e, events, places) : e.Location) is { } place && !string.IsNullOrWhiteSpace(place)) return place;
        return places.GetValueOrDefault("home");
    }

    static readonly Regex Postcode = new(@"\b([A-Z]{1,2}\d[A-Z\d]?)\s*(\d[A-Z]{2})\b", RegexOptions.IgnoreCase);
    static readonly Regex Coords = new(@"^\s*(-?\d{1,3}\.\d+)\s*,\s*(-?\d{1,3}\.\d+)\s*$");

    /// What TfL can plan from: a postcode without its space, or "lat,lon". Null for anything else (free text gets poor guesses).
    public static string? Code(string place)
    {
        if (Coords.Match(place) is { Success: true } c) return $"{c.Groups[1]},{c.Groups[2]}";
        return Postcode.Matches(place).LastOrDefault() is { } m ? (m.Groups[1].Value + m.Groups[2].Value).ToUpperInvariant() : null;
    }

    /// The one to catch: the last that still arrives by [by].
    public static Journey? Catch(IEnumerable<Journey> options, DateTime by) => options.Where(j => j.Arrive <= by).MaxBy(j => j.Depart);

    public static List<Journey> Merge(IEnumerable<Journey> a, IEnumerable<Journey> b) => a.Concat(b).DistinctBy(j => (j.Depart, j.Via)).OrderBy(j => j.Depart).ToList();

    /// TfL's journeys, plus the first leg's start and the last leg's end as "lat,lon" for the Citymapper link.
    public static (List<Journey> Journeys, string? Start, string? End) Parse(JsonNode j)
    {
        static DateTime T(JsonNode? n) => DateTime.Parse((string)n!, CultureInfo.InvariantCulture);
        static string? Point(JsonNode? p) => p?["lat"] is { } lat && p["lon"] is { } lon
            ? $"{((double)lat).ToString(CultureInfo.InvariantCulture)},{((double)lon).ToString(CultureInfo.InvariantCulture)}" : null;
        static string Line(JsonNode l)
        {
            var mode = (string?)l["mode"]?["id"];
            var name = (string?)l["routeOptions"]?[0]?["name"];
            if (string.IsNullOrEmpty(name)) return (string?)l["mode"]?["name"] ?? "?";
            return mode switch { "bus" => $"{name} bus", "tube" => $"{name} line", _ => name };
        }
        var list = new List<Journey>();
        string? start = null, end = null;
        foreach (var x in j["journeys"]?.AsArray() ?? new JsonArray())
        {
            var legs = x!["legs"]!.AsArray();
            var via = string.Join(" → ", legs.Where(l => (string?)l!["mode"]?["id"] != "walking").Select(l => Line(l!)));
            list.Add(new(T(x["startDateTime"]), T(x["arrivalDateTime"]), via == "" ? "Walk" : via));
            start ??= Point(legs[0]!["departurePoint"]);
            end ??= Point(legs[^1]!["arrivalPoint"]);
        }
        return (list, start, end);
    }

    static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(15), DefaultRequestHeaders = { { "User-Agent", "Glance" } } };

    /// Journeys arriving by (or leaving after) [at]. Empty when TfL finds no journey between the two.
    public static async Task<(List<Journey> Journeys, string? Start, string? End)> Fetch(string from, string to, DateTime at, bool arriving)
    {
        var url = $"https://api.tfl.gov.uk/Journey/JourneyResults/{Uri.EscapeDataString(from)}/to/{Uri.EscapeDataString(to)}" +
                  $"?date={at:yyyyMMdd}&time={at:HHmm}&timeIs={(arriving ? "Arriving" : "Departing")}";
        using var r = await Http.GetAsync(url);
        if (r.StatusCode is HttpStatusCode.MultipleChoices or HttpStatusCode.NotFound) return (new(), null, null);   // couldn't place an end, or no route
        r.EnsureSuccessStatusCode();
        return Parse(JsonNode.Parse(await r.Content.ReadAsStringAsync())!);
    }

    static string E(string s) => Uri.EscapeDataString(s);

    public static string Citymapper(string? start, string end, string endAddress, DateTime by) =>
        "https://citymapper.com/directions?" + (start != null ? $"startcoord={start}&" : "") +
        $"endcoord={end}&endaddress={E(endAddress)}&arrival_time={E(by.ToString("yyyy-MM-ddTHH:mm:sszzz", CultureInfo.InvariantCulture))}";

    public static string GoogleMaps(string from, string to) =>
        $"https://www.google.com/maps/dir/?api=1&origin={E(from)}&destination={E(to)}&travelmode=transit";
}

File: android/app/src/main/java/io/github/bosthecoder/glance/Travel.kt
package io.github.bosthecoder.glance

import java.net.URLEncoder
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

// Pure Kotlin like Model.kt: the rules in docs/travel.md, so TravelTest runs on the plain JVM.

/** One public transport option from TfL: [depart] already includes the walk to the stop. */
data class Journey(val depart: Long, val arrive: Long, val via: String)

object Travel {
    /** A timed event whose title starts with "Travel": "Travel: to Office", "Travel Home". */
    fun isTravel(e: Ev) = !e.allDay && e.title.trimStart().startsWith("travel", ignoreCase = true)

    /** The place a travel title names, and whether it's "from" it: "Travel: from Gym" -> ("gym", true). Work and office are one place. */
    fun name(title: String): Pair<String, Boolean> {
        var s = title.trim().drop(6).trimStart(':', '-', '–', ' ').trim().lowercase()
        val from = s.startsWith("from ")
        s = s.removePrefix("from ").removePrefix("to ").trim()
        return (if (s == "work") "office" else s) to from
    }

    /** Named places learned from travel events that have a location: "Travel: Home" at 35 Starboard Way teaches "home". */
    fun places(events: List<Ev>): Map<String, String> = events.filter { isTravel(it) && !it.location.isNullOrBlank() }
        .sortedBy { it.begin }
        .mapNotNull { e -> name(e.title).takeIf { !it.second && it.first.isNotEmpty() }?.let { it.first to e.location!!.trim() } }
        .toMap()

    private fun sameDay(a: Long, b: Long, zone: ZoneId) =
        Instant.ofEpochMilli(a).atZone(zone).toLocalDate() == Instant.ofEpochMilli(b).atZone(zone).toLocalDate()

    /** Where trip [t] goes: its location, a learned place, back to where "from X"'s outbound trip set off, or the next event's location. */
    fun destination(t: Ev, events: List<Ev>, places: Map<String, String>, zone: ZoneId): String? {
        t.location?.takeIf { it.isNotBlank() }?.let { return it.trim() }
        val (n, from) = name(t.title)
        if (!from) places[n]?.let { return it }
        if (from) events.filter { isTravel(it) && it.end <= t.begin && sameDay(it.begin, t.begin, zone) && name(it.title) == (n to false) }
            .maxByOrNull { it.begin }?.let { out -> origin(out, events, places, zone, null, 0)?.let { return it } }
        return events.filter { !it.allDay && !isTravel(it) && it.begin >= t.end && it.begin <= t.end + 120 * MIN && !it.location.isNullOrBlank() }
            .minByOrNull { it.begin }?.location?.trim()
    }

    /**
     * Where trip [t] starts: the device's [here] ("lat,lon") for a trip starting within 90 min of [now], the "from X"
     * place, the place of the last event before it that day (a trip's destination, another event's location), or home.
     */
    fun origin(t: Ev, events: List<Ev>, places: Map<String, String>, zone: ZoneId, here: String?, now: Long): String? {
        if (here != null && t.begin - now <= 90 * MIN) return here
        val (n, from) = name(t.title)
        if (from) places[n]?.let { return it }
        events.filter { !it.allDay && it !== t && it.end <= t.begin && sameDay(it.begin, t.begin, zone) }
            .sortedByDescending { it.end }
            .firstNotNullOfOrNull { e -> if (isTravel(e)) destination(e, events, places, zone) else e.location?.takeIf { it.isNotBlank() }?.trim() }
            ?.let { return it }
        return places["home"]
    }

    private val postcode = Regex("""\b([A-Z]{1,2}[0-9][A-Z0-9]?) ?([0-9][A-Z]{2})\b""", RegexOption.IGNORE_CASE)
    private val coords = Regex("""^\s*(-?\d{1,3}\.\d+)\s*,\s*(-?\d{1,3}\.\d+)\s*$""")

    /** What TfL takes for a place: a postcode without its space ("E162NZ") or "lat,lon". Null for anything else. */
    fun code(place: String): String? {
        coords.find(place)?.let { return "${it.groupValues[1]},${it.groupValues[2]}" }
        return postcode.findAll(place).lastOrNull()?.let { (it.groupValues[1] + it.groupValues[2]).uppercase() }
    }

    /** The options still worth showing: leave time (departure minus [bufferMin]) not yet passed, earliest first. */
    fun options(journeys: List<Journey>, now: Long, bufferMin: Int) =
        journeys.filter { it.depart - bufferMin * MIN >= now }.distinctBy { it.depart }.sortedBy { it.depart }

    /** The one to catch: the last option that still arrives by [arriveBy]. */
    fun catch(options: List<Journey>, arriveBy: Long) = options.lastOrNull { it.arrive <= arriveBy }

    private fun enc(s: String) = URLEncoder.encode(s, "UTF-8")

    fun citymapper(start: DoubleArray, end: DoubleArray, endAddress: String, arriveBy: Long, zone: ZoneId) =
        "https://citymapper.com/directions?startcoord=${start[0]},${start[1]}&endcoord=${end[0]},${end[1]}" +
            "&endaddress=${enc(endAddress)}&arrival_time=${enc(DateTimeFormatter.ISO_OFFSET_DATE_TIME.format(Instant.ofEpochMilli(arriveBy).atZone(zone).toOffsetDateTime()))}"

    fun googleMaps(from: String, to: String) =
        "https://www.google.com/maps/dir/?api=1&origin=${enc(from)}&destination=${enc(to)}&travelmode=transit"
}
